tests: Add Pannotia GPU Tests

erin-le opened the pull request:
This PR adds the Pannotia GPU tests. The input graphs to the binaries were too large to be committed, so they need to be obtained separately. Instructions for how to do so are in README.md.

weekly-tests.yaml was modified to download the input files, but I am not sure if it will work.

erin-le committed:
tests: Add Pannotia GPU tests

This commit adds the SE mode Pannotia GPU tests. Please note that
the input graphs were too large to be added, so they need to be obtained
separately.

Change-Id: I7b2a603ab08c94437546bf9d3125cf10172868b9

erin-le committed:
tests: update readme to include instructions to get inputs

Change-Id: I6b3a5359d558f1f6d6b3bf8b7495ad6e038bc64b

erin-le committed:
tests: Modify weekly-tests.yaml to download Pannotia inputs

Change-Id: If47412410dd9c845641469822ac46aabf24310a7

mattsinc commented:
Thanks for doing this, @erin-le !  Let me know when you want me to review this -- is it WIP?

abmerop commented:
This looks fine to me at a quick glance.  Brad says I should not spend time on SE mode GPU stuff so I will defer mostly to @mattsinc  :) . 

Is it possible yet to do the same tests for FS mode?  I believe we have fixed a few issues with KVM on develop that would help but that should probably be discussed in #787 and #456 not here.

BobbyRBruce commented on tests/gem5/gpu-pannotia-resources/pannotia-bins/color_maxmin.gem5:
You accidentally added this directory. At the very least you've got binaries here you're also `wget`ing in the script.

For the datasets, ``wget them too. Binaries and raw data doesn't play nice with git.

BobbyRBruce commented on tests/gem5/gpu/README.md:
Looking over this file: little of it makes sense (that main.py command isn't even using the gem5/gpu directory).

BobbyRBruce commented on tests/gem5/gpu/README.md:
````diff
@@ -6,3 +6,18 @@ To run these tests by themselves, you can run the following command in the tests
+To obtain the input graphs needed to run the Pannotia tests, run the following
+in tests/gpu-pannotia-resources/pannotia-datasets:
+
+```bash
+gsutil -m cp -r \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/README.txt" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/bc" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/color" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/floydwarshall" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/mis" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/pagerank" \
+"gs://dist.gem5.org/dist/develop/datasets/pannotia/sssp" \
+.
+```
````
We shouldnt be having the user do this. Just have the python script wget the resources was needed.

Let's do this the cheap way: put the wget's (or whatever the python equivalent is) in tests_gpu_pannotia.py. Just have a little part at the top of the file that checks if tests/gem5/resources/pannotia exits, and, if not, creates the directory downloads all the binaries and datafiles into it.

BobbyRBruce commented on .github/workflows/weekly-tests.yaml:
```diff
@@ -71,7 +71,11 @@ jobs:
-
+            - name: Get Pannotia inputs
+              working-directory: ${{ github.workspace }}/tests/gem5/gpu-pannotia-resources/pannotia-datasets
+              run: gsutil -m cp -r \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/README.txt" \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/bc"
+                  \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/color" \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/floydwarshall" \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/mis"
+                  \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/pagerank" \ "gs://dist.gem5.org/dist/develop/datasets/pannotia/sssp" \ .
```
`gsutil` is mostly a tool for people to interact with their Google Cloud account. I think you need to associate a Google Account. Anyway, I doubt it's installed on our runners and it's really not the tool for the job. If we were to put the download here i'd just recommend `wget`, though I have a suggestion elsewhere in this review that we should put them in "tests_gpu_pannotia.py"

erin-le committed:
tests: get binaries and input files for GPU tests in test script

This commit makes it so that the testlib script
automatically downloads the binaries and input datasets, instead
of making the user do it or having the download command in the
workflow files.

erin-le committed:
Merge branch 'develop' into weekly-tests-gpu-tests

mattsinc commented:
Thanks @BobbyRBruce for taking a pass on this.  @erin-le let me know when you'd like me to review it -- or if it's still WIP.

erin-le committed:
tests: change hardcoding, minor wording adjustment

erin-le commented:
@mattsinc Sorry for the delay; it's ready for review now.

mattsinc commented on tests/gem5/gpu/test_gpu_pannotia.py:
```diff
@@ -0,0 +1,344 @@
+    length=constants.very_long_tag,
+)
+
+gem5_verify_config(
```
Nothing wrong with this, but previously we only ran 1 input set per app in the weekly tests.  Running more is fine if we have time/resources, but technically speaking this is not necessary.  Open to discussion though.

mattsinc commented on tests/gem5/gpu/test_gpu_pannotia.py:
```diff
@@ -0,0 +1,344 @@
+    length=constants.very_long_tag,
+)
+
+gem5_verify_config(
```
Previously for color we were running smaller inputs (https://github.com/gem5/gem5/blob/stable/tests/deprecated/weekly.sh#L275) to make the tests complete in a more tractable time.  I believe these test sizes you have will likely run for multiple days.  If that is ok, then fine but just an FYI.

mattsinc commented on tests/gem5/gpu/test_gpu_pannotia.py:
```diff
@@ -0,0 +1,344 @@
+    length=constants.very_long_tag,
+)
+
+gem5_verify_config(
```
same for FW: https://github.com/gem5/gem5/blob/stable/tests/deprecated/weekly.sh#L309

mattsinc commented on tests/gem5/gpu/test_gpu_pannotia.py:
```diff
@@ -0,0 +1,344 @@
+    length=constants.very_long_tag,
+)
+
+# This test fails with
```
I have a fix for this locally but was not aware it was failing for people in public -- I thought it was only for the multi-chiplet support (not on mainline).

Same idea for MIS though about input size: https://github.com/gem5/gem5/blob/stable/tests/deprecated/weekly.sh#L324

mattsinc commented on tests/gem5/gpu/test_gpu_pannotia.py:
```diff
@@ -0,0 +1,344 @@
+    length=constants.very_long_tag,
+)
+
+gem5_verify_config(
```
Same with SSSP: https://github.com/gem5/gem5/blob/stable/tests/deprecated/weekly.sh#L371

mattsinc reviewed (commented):
I realize this has already been merged in, but I wanted to add a few comments on things here, @erin-le .  Thanks again for all your hard work and help with this.  I think the main question is if we should run such large input sizes for Pannotia -- the test coverage is more than the smaller ones, but at very large increases in runtime (and running out of memory in some of the cases, as you have documented).

Happy to push a follow-up that makes those changes or let you do so.

mattsinc commented:
@erin-le just wanted to make sure you saw these comments

I think some of what @BobbyRBruce is seeing in #1628 is related to these comments

erin-le commented:
@mattsinc: The dev team has seen these comments and @Harshil2107 is currently working on making the changes

Harshil2107 commented:
@mattsinc: I pushed a PR #1631 with the changes you requested. I am still running the tests locally to see how long they take and if there are any errors.
